src/main/java/dev/ecattez/shahmat/board/ChessGame.java

```diff
@@ -0,0 +1,188 @@
+package dev.ecattez.shahmat.board;
+
+import dev.ecattez.shahmat.board.move.Capture;
+import dev.ecattez.shahmat.board.move.EnPassant;
+import dev.ecattez.shahmat.board.move.Movement;
+import dev.ecattez.shahmat.board.move.MovementVisitor;
+import dev.ecattez.shahmat.board.move.MovingStrategy;
+import dev.ecattez.shahmat.board.move.MoveOnVacant;
+import dev.ecattez.shahmat.board.pawn.PawnPromotionAllowedPieces;
+import dev.ecattez.shahmat.event.PromotionProposed;
+import dev.ecattez.shahmat.board.pawn.PawnPromotionRankVisitor;
+import dev.ecattez.shahmat.command.Move;
+import dev.ecattez.shahmat.command.Promote;
+import dev.ecattez.shahmat.event.BoardEvent;
+import dev.ecattez.shahmat.event.PawnPromoted;
+import dev.ecattez.shahmat.event.PieceCaptured;
+import dev.ecattez.shahmat.event.PieceMoved;
+import dev.ecattez.shahmat.event.PiecePositioned;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class ChessGame {
+
+    private static final PieceFactory PIECE_FACTORY = new PieceBox(); // fixme: change it
+
+    private static final PieceTypeVisitor<MovingStrategy> MOVING_RULES = new MovingRules();
+
+    private static final PieceTypeVisitor<Boolean> PIECES_THAT_CAN_PROMOTE = new PawnPromotionAllowedPieces();
+
+    private static final PieceColorVisitor<Integer> PROMOTION_RANKS = new PawnPromotionRankVisitor();
+
+    private static final EventMovementVisitor EVENTS_FROM_MOVEMENT = new EventMovementVisitor();
+
+    public static List<BoardEvent> move(
+        List<BoardEvent> history,
+        Move command
+    ) throws RulesViolation {
+        Square from = command.from;
+        Square to = command.to;
+
+        if (from.equals(to)) {
+            throw new InvalidMove("Current position and destination are the same square");
+        }
+
+        Board board = replay(history);
+        if (board.isVacant(from)) {
+            throw new NoPieceOnSquare(from);
+        }
+
+        Piece pieceToMove = board
+            .getPiece(from)
+            .filter(command.piece::equals)
+            .orElseThrow(() -> new WrongPieceSelected(from));
+
+        List<BoardEvent> events = getMovement(pieceToMove, from, to, board, history)
+            .stream()
+            .map(move -> move.accept(EVENTS_FROM_MOVEMENT))
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+
+        if (events.isEmpty()) {
+            throw new ImpossibleToMove(pieceToMove, from, to);
+        }
+
+        if (canBePromoted(pieceToMove, to)) {
+            events.add(
+                new PromotionProposed(to)
+            );
+        }
+
+        return events;
+    }
+
+    public static List<BoardEvent> promote(
```

Comment:
Same thing as the `move` method: Maybe make this method a composition of private ones ?